File: client-clojure/src/sirqul_io_t_platform/specs/purchase_item_response.clj
(ns sirqul-io-t-platform.specs.purchase-item-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            )
  (:import (java.io File)))


(def purchase-item-response-data
  {
   (ds/opt :id) int?
   (ds/opt :name) string?
   (ds/opt :description) string?
   (ds/opt :tickets) int?
   (ds/opt :coverAsset) asset-short-response-spec
   (ds/opt :promoAsset) asset-short-response-spec
   (ds/opt :giftable) boolean?
   (ds/opt :assetable) boolean?
   (ds/opt :gameType) string?
   (ds/opt :price) float?
   (ds/opt :purchaseType) string?
   (ds/opt :purchaseCode) string?
   (ds/opt :purchaseLimit) int?
   (ds/opt :quantity) int?
   (ds/opt :serviceAction) string?
   (ds/opt :appName) string?
   })

(def purchase-item-response-spec
  (ds/spec
    {:name ::purchase-item-response
     :spec purchase-item-response-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/PurchaseItemResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import PurchaseItemResponse._

case class PurchaseItemResponse (
  id: Option[Long],
name: Option[String],
description: Option[String],
tickets: Option[Integer],
coverAsset: Option[AssetShortResponse],
promoAsset: Option[AssetShortResponse],
giftable: Option[Boolean],
assetable: Option[Boolean],
gameType: Option[String],
price: Option[Float],
purchaseType: Option[String],
purchaseCode: Option[String],
purchaseLimit: Option[Integer],
quantity: Option[Integer],
serviceAction: Option[ServiceAction],
appName: Option[String])

object PurchaseItemResponse {
  import DateTimeCodecs._
  sealed trait ServiceAction
  case object DAYPREMIUM extends ServiceAction
  case object WEEKPREMIUM extends ServiceAction
  case object MONTHPREMIUM extends ServiceAction
  case object YEARPREMIUM extends ServiceAction
  case object LIFETIMEPREMIUM extends ServiceAction
  case object ADDTICKET extends ServiceAction
  case object ADDGIFT extends ServiceAction
  case object ADDPOINTS extends ServiceAction

  object ServiceAction {
    def toServiceAction(s: String): Option[ServiceAction] = s match {
      case "DAYPREMIUM" => Some(DAYPREMIUM)
      case "WEEKPREMIUM" => Some(WEEKPREMIUM)
      case "MONTHPREMIUM" => Some(MONTHPREMIUM)
      case "YEARPREMIUM" => Some(YEARPREMIUM)
      case "LIFETIMEPREMIUM" => Some(LIFETIMEPREMIUM)
      case "ADDTICKET" => Some(ADDTICKET)
      case "ADDGIFT" => Some(ADDGIFT)
      case "ADDPOINTS" => Some(ADDPOINTS)
      case _ => None
    }

    def fromServiceAction(x: ServiceAction): String = x match {
      case DAYPREMIUM => "DAYPREMIUM"
      case WEEKPREMIUM => "WEEKPREMIUM"
      case MONTHPREMIUM => "MONTHPREMIUM"
      case YEARPREMIUM => "YEARPREMIUM"
      case LIFETIMEPREMIUM => "LIFETIMEPREMIUM"
      case ADDTICKET => "ADDTICKET"
      case ADDGIFT => "ADDGIFT"
      case ADDPOINTS => "ADDPOINTS"
    }
  }

  implicit val ServiceActionEnumEncoder: EncodeJson[ServiceAction] =
    EncodeJson[ServiceAction](is => StringEncodeJson(ServiceAction.fromServiceAction(is)))

  implicit val ServiceActionEnumDecoder: DecodeJson[ServiceAction] =
    DecodeJson.optionDecoder[ServiceAction](n => n.string.flatMap(jStr => ServiceAction.toServiceAction(jStr)), "ServiceAction failed to de-serialize")

  implicit val PurchaseItemResponseCodecJson: CodecJson[PurchaseItemResponse] = CodecJson.derive[PurchaseItemResponse]
  implicit val PurchaseItemResponseDecoder: EntityDecoder[PurchaseItemResponse] = jsonOf[PurchaseItemResponse]
  implicit val PurchaseItemResponseEncoder: EntityEncoder[PurchaseItemResponse] = jsonEncoderOf[PurchaseItemResponse]
}
